engine/table/src/main/java/io/deephaven/engine/util/SourceClosure.java

```diff
@@ -16,6 +16,7 @@
  * This class represents uses of the source() and sourceOnce() method calls. It will use the underlying scriptPathLoader
  * to get the script text to use.
  */
+@Deprecated(forRemoval = true)
```

Comment:
This is an enterprise thing, but I think it might be best to delete it. We should discuss with @cpwright .

Reply:
You are using it too from what I can tell.
        setVariable("source", sourceClosure = new SourceClosure(this, pathLoader, false, caching));
        setVariable("sourceOnce", sourceOnceClosure = new SourceClosure(this, pathLoader, true, false));

I think that Groovy source is kind of terrible in a lot of ways.  It is our own invention and has some unpleasant semantics when you mix the code between the sourced thing and the rest of the code.  I don't know if we can do something better given the natural language constructs, because I am not a Groovy expert.

I think Raffi and Pete might want to kick us in the shins if we take it away without a suitable replacement though.

Reply:
> You are using it too from what I can tell.

That would be true if setScriptPathLoader was ever called - there is one callsite, from GroovyDeephavenSession.onApplicationInitializationBegin, which itself is never called, so we stick with the default impl in groovy:

```
source = {fileName ->
    __groovySession.runScript(fileName)
}

sourceOnce = {fileName ->
    __groovySession.runScriptOnce(fileName)
}
```

After some discussion with Charles and Raffi, I'm going to create a DHC and DHE issue to track this further, but the simple consensus is that we're going to deprecate/remove source and sourceOnce, and instead encourage using Groovy's own language features for this sort of thing, like `import`, and potentially a custom classloader to find resources not on the usual classpath (i.e. a script/notebook directory, etc). Re-adding source/sourceOnce (or something like it) could be done in a given installation at the discretion of local admins, but shouldn't be strictly necessary to use Groovy as it is "intended" to be used.

https://github.com/deephaven/deephaven-core/issues/4160